### ISSUE TITLE
Ability to return an iterator on a "many" query

### ISSUE THREAD

**Issue by encendre** [kyleconroy/sqlc#720]
Hi, when a query return a myriad of objects, the slice based solution is not very efficient, it would be great if it was possible to have a way to return some kind of iterator on the rows, something like sql.Rows but type safe.

The first solution that comes to mind is to add a new type of function in the sql queries builder.
Instead of writing `-- name: Name :many` we would write `-- name: Name :rows` or another relevant keyword

and for the generated code, it would look something like this
```sql
CREATE TABLE pilots (
          id integer NOT NULL,
          name text NOT NULL
);

ALTER TABLE pilots ADD CONSTRAINT pilot_pkey PRIMARY KEY (id);
```
```sql
-- name: ListPilots :rows
SELECT * FROM pilots;
```
```go
type Pilot struct {
	ID   int32
	Name string
}

type PilotRows struct {
	rows *sql.Rows
}

func (p *PilotRows) Close() error {
	return p.rows.Close()
}
func (p *PilotRows) Next() bool {
	return p.rows.Next()
}
func (p *PilotRows) Scan(dest *Pilot) error {
	return p.rows.Scan(dest.ID, dest.Name)
}
func (p *PilotRows) Err() error {
	return p.rows.Err()
}

func (q *Queries) ListPilots(ctx context.Context) (*PilotRows, error) {
	rows, err := q.query(ctx, q.listPilotsStmtn, listPilots)
	if err != nil {
		return nil, err
	}
	return &PilotRows{rows: rows}, nil
}

```

**Comment by kyleconroy** [comment]
@encendre thanks for the feature request. You're suggestion fixes a few of the issues we discussed when this first came up in #155. I'd ask you the same question I asked in that thread: can you use some type of paging to work through this query? It's not usually a good idea to request thousands of rows from a single query.

**Comment by eivindr** [comment]
@encendre @kyleconroy In #155 I linked to a poc for streaming via an iterator func. It's used in a project that needs to stream out alot of entries. Paging should oc be preferred, but sometimes its not an option.

**Comment by pcriv** [comment]
My team and I are considering using sqlc for our services and having this feature would be a huge pro for going for sqlc. Anyway, we could help with this?

**Comment by pcriv** [comment]
@eivindr Did you find a nice way to extend sqlc to support the iterator?

**Comment by eivindr** [comment]
@pcriv I changed the generator to produce code like https://github.com/kyleconroy/sqlc/issues/155#issuecomment-653754008 - so you pass in a iterator func that handles the streaming. If the func returns error the stream is aborted. You tell me if its a nice way or not :P

**Comment by pcriv** [comment]
That one looks good! but I assume you are using your fork for now. I was hoping there was a nice way of extending sqlc without having to rely on a fork 😅

**Comment by eivindr** [comment]
Yes, I'm just using the fork. I could create a PR if @kyleconroy pre-approves the fundamentals in the approach. The PR would also need to have some documentation, and maybe also generate similar for Kotlin..? :sweat_smile:

**Comment by pcriv** [comment]
Having this in sqlc would be a huge win for us haha, If I can help somehow please let me know

**Comment by pcriv** [comment]
@eivindr maybe you can just open the PR and we can continue the discussion there?

**Comment by kyleconroy** [comment]
We're planning on adding support for this if the "range over func" proposal is accepted: https://github.com/golang/go/issues/61405

**Comment by dskarataev** [comment]
Hi @kyleconroy , the mentioned proposal has been accepted already. Would it be possible to go ahead with the requested feature? We also really need it for the case of exporting data in the streaming mode and where it is not convenient to chunk the data by running multiple queries.

Thanks a lot in advance!

**Comment by kyleconroy** [comment]
Range over func will be enabled by default on Go 1.23, so we're now unblocked from the Go side of things.

**Comment by tgmendes** [comment]
Hi @kyleconroy 👋 

We're also quite interested in start using the range over func for these queries - do you know if they'll be available any time soon? Or if there's a workaround to configure the generator to use these?

Thanks!

**Comment by viewsharp** [comment]
Hi @kyleconroy 
Can I help you with implementation of this feature?

In my opinion, there are 2 main ways to implement range-over-func. Which one would you prefer?
Option 1:

```go
func (q *Queries) IterCities(ctx context.Context) (iter.Seq[City], func() error) {
	var err error
	errFunc := func() error {
		return err
	}

	rows, err := q.query(ctx, q.listCitiesStmt, iterCities)
	if err != nil {
		return func(yield func(City) bool) {}, errFunc
	}

	return func(yield func(City) bool) {
		defer rows.Close()

		for rows.Next() {
			var i City
			err = rows.Scan(&i.Slug, &i.Name)
			if err != nil {
				return
			}

			if !yield(i) {
				return
			}
		}

		err = rows.Close()
		if err != nil {
			return
		}
		err = rows.Err()
	}, errFunc
}

func example() {
	iter, errFunc := q.IterCities(ctx)
	for city := range iter {
		// do something with city
	}
	if err := errFunc(); err != nil {
		// handle error
	}
}
```

Option 2:
```go
func (q *Queries) IterCities2(ctx context.Context) iter.Seq2[City, error] {
	rows, err := q.query(ctx, q.listCitiesStmt, iterCities)
	if err != nil {
		return func(yield func(City, error) bool) {
			yield(City{}, err)
		}
	}

	return func(yield func(City, error) bool) {
		defer rows.Close()

		for rows.Next() {
			var i City
			err = rows.Scan(&i.Slug, &i.Name)
			if !yield(i, err) {
				return
			}
		}

		err = rows.Close()
		if err != nil {
			yield(City{}, err)
			return
		}

		err = rows.Err()
		if err != nil {
			yield(City{}, err)
		}
	}
}

func example2() {
	for city, err := range q.IterCities2(ctx) {
		if err != nil {
			// handle error
			break
		}
		// do something with city
	}
}
```

**Comment by kyleconroy** [comment]
I found this article really helpful in understanding how to range over functions in Go 1.23 https://eli.thegreenplace.net/2024/ranging-over-functions-in-go-123/

**Comment by crhntr** [comment]
I haven't seen this pattern anywhere but passing a handler into the iter.Seq constructor seems like it aligns with the spirit of iterators. https://go.dev/play/p/ZVQ1-pwNruC

```txtar
package main

import (
	"fmt"

	"play/iterate"
)

func main() {
	for n := range iterate.Numbers(5, func(err error) bool {
		fmt.Println("Error:", err)
		return false
	}) {
		fmt.Println(n)
	}
}

-- iterate/numbers.go --
package iterate

import (
	"errors"
	"iter"
)

var Err3 = errors.New("3 is a crowd")

func Numbers(n int, handle func(error) bool) iter.Seq[int] {
	return func(yield func(int) bool) {
		for i := range n {
			if i == 3 {
				if !handle(Err3) {
					return
				}
			}
			if !yield(i) {
				return
			}
		}
	}
}
```

**Comment by MatthiasKunnen** [comment]
I am using the following in production at the moment:

## How to use
```go
results, err := db.Get(ctx)
if err != nil {
	return err
}

for result, rangeErr := range results {
	if rangeErr != nil {
		return rangeErr
	}
	
	fmt.Println("%v\n", result)
}

return nil
```

## Generated code
```go
type Result struct {
	Id     uuid.UUID
	Domain string
}

func (q *Queries) Get(ctx context.Context) (iter.Seq2[*Result error], error) {
	rows, err := q.db.Query(ctx, query)
	if err != nil {
		return nil, err
	}

	return func(yield func(*Result, error) bool) {
		defer rows.Close()

		for {
			// @todo .Next and .Scan do not stop on context cancellation, is this expected of pgx?
			select {
			case <-ctx.Done():
				yield(nil, ctx.Err())
				return
			default:
			}

			if !rows.Next() {
				if returnErr := rows.Err(); returnErr != nil {
					yield(nil, returnErr)
				}
				return
			}

			var id pgtype.UUID
			var domain pgtype.Text
			if scanErr := rows.Scan(&id, &domain); scanErr != nil {
				yield(nil, scanErr)
				return
			}

			if !yield(&Result{
				Id:     id.Bytes,
				Domain: domain.String,
			}, nil) {
				return
			}
		}
	}, nil
}
```

I still have to look into whether context cancellation not being handled by pgx is expected.

**Comment by e-gloo** [comment]
> Range over func will be enabled by default on Go 1.23, so we're now unblocked from the Go side of things.

Any updates regarding this feature, since there's no limitations from Go now?

**Comment by Tilzen** [comment]
## Proposal: opt-in iterator generation for `:many` queries (reviving #720)

Following up on the [Feb 2026 ask](https://github.com/sqlc-dev/sqlc/issues/720#issuecomment-...) for updates, and cross-linking [#4464](https://github.com/sqlc-dev/sqlc/issues/4464) — there is **no open PR** for this today ([#3631](https://github.com/sqlc-dev/sqlc/pull/3631) was closed without merge). Go 1.23+ iterators are stable; the remaining gap is an agreed API and config surface.

Below is a consolidated proposal that tries to resolve the open design threads from #3631 while keeping **zero breaking changes**.

---

### Problem

sqlc's `:many` queries always materialize into `[]T`. That is great DX for small result sets, but for exports, sync jobs, ETL, and backfills it forces **O(n) heap allocation** even when callers only need row-by-row processing. Workarounds (manual paging, forks, skipping sqlc on hot paths) all sacrifice type safety or maintenance.

Go 1.23 shipped `iter.Seq` / `iter.Seq2`. Kyle [noted](https://github.com/sqlc-dev/sqlc/issues/720#issuecomment-...) that this unblocks native iterator generation. The community need is still very much alive ([#720](https://github.com/sqlc-dev/sqlc/issues/720), [#4464](https://github.com/sqlc-dev/sqlc/issues/4464), production patterns shared by [@MatthiasKunnen](https://github.com/sqlc-dev/sqlc/issues/720#issuecomment-...)).

---

### Design principles

1. **Do not change default `:many` behavior** — `ListAuthors(ctx) ([]Author, error)` stays as-is.
2. **Opt-in via config** (and optionally per-query annotation).
3. **Go 1.23 native idioms** — primary output is `iter.Seq2[T, error]`.
4. **Lazy by default** — the DB query runs on the first iteration step, not at method call (avoids dangling queries if the iterator is never consumed).
5. **Correct lifecycle** — `defer rows.Close()` inside the iterator closure (addresses [gbarr's concern in #3631](https://github.com/sqlc-dev/sqlc/pull/3631#issuecomment-...)).

---

### Proposed `sqlc.yaml` options

```yaml
gen:
  go:
    emit_iterators: false          # opt-in; default false

    iterator_scope: global         # global | explicit_only
    iterator_method_prefix: "Iter" # ListAuthors → IterAuthors (or "Stream")
    iterator_style: seq2             # seq2 | callback | rows
    iterator_start: lazy             # lazy | eager
```

**Scope modes:**

| Mode | Behavior |
|------|----------|
| `emit_iterators: true` + `iterator_scope: global` | Every `:many` query gets a streaming companion |
| `iterator_scope: explicit_only` | Only annotated queries, e.g. `-- name: ListAuthors :many:stream` or `-- name: StreamAuthors :stream` ([#4464](https://github.com/sqlc-dev/sqlc/issues/4464)) |

Teams that want simple DX flip one flag; teams that want fine-grained control use annotations.

---

### Generated API (recommended default: `seq2` + `lazy`)

**SQL — unchanged:**

```sql
-- name: ListAuthors :many
SELECT id, name, bio FROM authors ORDER BY name;
```

**Go — existing + new companion:**

```go
// unchanged
func (q *Queries) ListAuthors(ctx context.Context) ([]Author, error) { ... }

// new (when emit_iterators: true)
func (q *Queries) IterAuthors(ctx context.Context) iter.Seq2[Author, error] {
    return func(yield func(Author, error) bool) {
        rows, err := q.db.QueryContext(ctx, listAuthors)
        if err != nil {
            yield(Author{}, err)
            return
        }
        defer rows.Close()

        for rows.Next() {
            var i Author
            if err := rows.Scan(&i.ID, &i.Name, &i.Bio); err != nil {
                yield(Author{}, err)
                return
            }
            if !yield(i, nil) {
                return
            }
        }
        if err := rows.Err(); err != nil {
            yield(Author{}, err)
        }
    }
}
```

**Usage:**

```go
for author, err := range q.IterAuthors(ctx) {
    if err != nil {
        return fmt.Errorf("list authors: %w", err)
    }
    if err := process(author); err != nil {
        return err
    }
}
```

This aligns with Kyle's [later preference in #3631](https://github.com/sqlc-dev/sqlc/pull/3631#issuecomment-...) for direct `for x, err := range q.Method(ctx)` — no wrapper type in the common case, hard to misuse, early `break` still closes rows via `defer`.

---

### Alternate styles (configurable, not all generated by default)

| `iterator_style` | Generated method | Notes |
|------------------|------------------|-------|
| `seq2` (default) | `IterAuthors(ctx) iter.Seq2[Author, error]` | Recommended |
| `callback` | `EachAuthor(ctx, func(Author) error) error` | Optional sugar; can also be a one-liner at call sites |
| `rows` | `*IterAuthorsRows` with `All()` / `Err()` / `Close()` | Compatibility with [#720](https://github.com/sqlc-dev/sqlc/issues/720) original request and #3631 |

**Not proposing three methods per query by default** — `List*` + `Iter*` covers slice vs stream; `Each*` is optional config.

**`iterator_start: eager`** (alternative): return `(iter.Seq2[T, error], error)` so connection errors surface before iteration — useful for some pgx setups, but lazy should be the default.

---

### Parameterized queries

Same pattern — iterators mirror `:many` signatures:

```sql
-- name: ListAuthorsByIDs :many
SELECT id, name, bio FROM authors WHERE id = ANY($1::int[]);
```

```go
func (q *Queries) ListAuthorsByIDs(ctx context.Context, ids []int32) ([]Author, error)
func (q *Queries) IterAuthorsByIDs(ctx context.Context, ids []int32) iter.Seq2[Author, error]
```

---

### Suggested v1 scope

- **Go** codegen only
- **`database/sql`** driver path first (pgx follow-up)
- **PostgreSQL** example + e2e test (pattern from #3631's examples)
- **Min Go 1.23+** when `emit_iterators: true` (document in release notes)
- Python/Kotlin: coordinate separately ([#4464](https://github.com/sqlc-dev/sqlc/issues/4464), [@borissmidt's offer](https://github.com/sqlc-dev/sqlc/issues/4464#issuecomment-...))

---

### Open questions for @kyleconroy / maintainers

1. **Method prefix:** `Iter` vs `Stream`?
2. **Lazy default:** agree lazy is correct for v1?
3. **Config vs annotation:** ship both `emit_iterators: global` and `:stream` explicit mode?
4. **Eager mode:** v1 or defer?
5. **pgx:** same PR or immediate follow-up?

---

### Offer to implement

Happy to put together a minimal PoC PR once there is sign-off on the config surface and default API (`seq2` + lazy + `Iter` prefix). Can start with one `:many` query, one engine, and an end-to-end test.

**One-liner:** sqlc already generates excellent type-safe Go; an opt-in `emit_iterators` flag would add lazy `iter.Seq2[T, error]` companions for `:many` — same type safety, O(1) memory per row, no breaking changes.

Related: [#4464](https://github.com/sqlc-dev/sqlc/issues/4464), [#4108](https://github.com/sqlc-dev/sqlc/issues/4108), closed reference [#3631](https://github.com/sqlc-dev/sqlc/pull/3631).

**Comment by Tilzen** [comment]
Opened a **proof-of-concept PR** (not merge-ready) implementing the proposal above for early feedback:

https://github.com/sqlc-dev/sqlc/pull/4488

PoC branch / reference: https://github.com/Tilzen/sqlc/tree/feat/emit-iterators-poc

Includes `emit_iterators` config, lazy `iter.Seq2[T, error]` companions for `:many`, `:stream` / `:many:stream` annotations, stdlib + pgx/v5 templates, PRD doc, and runtime + codegen tests. Looking for maintainer sign-off on API/config before polishing toward a production PR.

**Comment by sgielen** [comment]
For what it's worth, I've been using a iter.Seq heavy API library which uses this format:

```
objects, errf := c.GetObjects(ctx)
for object := range objects {
  // ....
}
if err := errf(); err != nil {
  // ...
}
```

In my experience working with this, this has been quite natural, any final GetObjects error is not inside the loop, and the Go linter gives an error when `errf` is not referenced (just like `err` is when not explicitly checked).

**Comment by iitsdani** [comment]
Agreed with @sgielen - the smart pointer approach also mimics what [`sql.Rows`](https://pkg.go.dev/database/sql#Rows) does (where `.Next()` is replaced by the `iter.Seq` pattern).
